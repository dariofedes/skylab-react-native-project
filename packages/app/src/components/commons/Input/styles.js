import { StyleSheet } from "react-native";

const styles = StyleSheet.create({
    input: {
        borderColor: "#333",
        borderWidth: 1,
        borderRadius: 10,
        paddingHorizontal: 10,
        paddingVertical: 10,
        color: '#333',
        margin: 10
    },
})

export default styles