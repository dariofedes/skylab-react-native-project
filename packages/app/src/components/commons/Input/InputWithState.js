import React, { useState } from 'react'
import { TextInput } from 'react-native'
import PropTypes from 'prop-types'

import { getKeyboardType } from '../../../utils/Device'
import styles from './styles'


// To test onChangeText method

const Input = ({ placeholder, type }) => {
    const isSecure = type === 'password'

    const [value, setValue] = useState('')

    return (
        <TextInput 
            style={styles.input}
            placeholder={placeholder}
            placeholderTextColor='#666'
            secureTextEntry={isSecure}
            value={value}
            onChangeText={(v) => setValue(v)}
            keyboardType={getKeyboardType(type)}
        />
    )
}

Input.propTypes = {
    placeholder: PropTypes.string,
    type: PropTypes.string
}

Input.defaultProps = {
    placeholder: 'Type here...',
    type: 'default'
}

export default Input