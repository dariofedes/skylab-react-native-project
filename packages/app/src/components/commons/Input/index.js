import React from 'react'
import { TextInput } from 'react-native'

import { getKeyboardType } from '../../../utils/Device'
import styles from './styles'

const Input = ({ placeholder, onChangeText, value, type = 'default' }) => {
    const isSecure = type === 'password'

    return (
        <TextInput 
            style={styles.input}
            placeholder={placeholder}
            placeholderTextColor='#666'
            secureTextEntry={isSecure}
            value={value}
            onChangeText={onChangeText}
            keyboardType={getKeyboardType(type)}
        />
    )
}

export default Input