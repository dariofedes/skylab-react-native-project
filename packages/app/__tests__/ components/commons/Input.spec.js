import React from 'react'
import Renderer from 'react-test-renderer'
import ShallowRenderer from 'react-test-renderer/shallow'
import { TextInput } from 'react-native'

import Input from '@skylab/app/src/components/commons/Input'

describe('Input Component', () => {
    describe('Validate props', () => {
        // it('throw error if it has not the required props', () => {
        //     expect(Renderer.create(<Input/>)).toThrow()
        // })

        it('render with default props', () => {
            const mockFunc = jest.fn()

            const renderer = Renderer.create(<Input onChangeText={mockFunc} value='value'/>)
            const output = renderer.root

            expect(output.props.placeholder).toStrictEqual('Type here...')
            expect(output.props.onChangeText).toStrictEqual(mockFunc)
            expect(output.props.value).toStrictEqual('value')
            expect(output.props.type).toStrictEqual('default')

            const rnInput = output.findByType(TextInput)
            
            expect(rnInput.props.secureTextEntry).toBeFalsy()
            expect(rnInput.props.keyboardType).toStrictEqual('default')
        })

        it('render with password and secure text entry', () => {
            const mockFunc = jest.fn()

            const renderer = Renderer.create(
                <Input
                    onChangeText={mockFunc}
                    value='value'
                    placeholder='placeholder'
                    type='password'
                />
            )
            const output = renderer.root

            expect(output.props.placeholder).toStrictEqual('placeholder')
            expect(output.props.onChangeText).toStrictEqual(mockFunc)
            expect(output.props.value).toStrictEqual('value')
            expect(output.props.type).toStrictEqual('password')

            const rnInput = output.findByType(TextInput)
            
            expect(rnInput.props.secureTextEntry).toBeTruthy()
            expect(rnInput.props.keyboardType).toStrictEqual('default')
        })

        it('render with phone type input', () => {
            const mockFunc = jest.fn()

            const renderer = Renderer.create(
                <Input
                    onChangeText={mockFunc}
                    value='value'
                    placeholder='placeholder'
                    type='phone'
                />
            )
            const output = renderer.root

            expect(output.props.placeholder).toStrictEqual('placeholder')
            expect(output.props.onChangeText).toStrictEqual(mockFunc)
            expect(output.props.value).toStrictEqual('value')
            expect(output.props.type).toStrictEqual('phone')

            const rnInput = output.findByType(TextInput)
            
            expect(rnInput.props.secureTextEntry).toBeFalsy()
            expect(rnInput.props.keyboardType).toStrictEqual('numeric')
        })

        it('render with email type input', () => {
            const mockFunc = jest.fn()

            const renderer = Renderer.create(
                <Input
                    onChangeText={mockFunc}
                    value='value'
                    placeholder='placeholder'
                    type='email'
                />
            )
            const output = renderer.root

            expect(output.props.placeholder).toStrictEqual('placeholder')
            expect(output.props.onChangeText).toStrictEqual(mockFunc)
            expect(output.props.value).toStrictEqual('value')
            expect(output.props.type).toStrictEqual('email')

            const rnInput = output.findByType(TextInput)
            
            expect(rnInput.props.secureTextEntry).toBeFalsy()
            expect(rnInput.props.keyboardType).toStrictEqual('email-address')
        })
    })

    // describe('Methods', () => {

    //      This should be a test for situations such as change input value in the same compo
    //      InputWithState Component

    //     it('should call onChangeText method correctly', () => {
    //         const renderer = Renderer.create(<Input/>)
    //         const output = renderer.root

    //         const rnInput = output.findByType(TextInput)

    //         expect(rnInput.props.value).toStrictEqual('')

    //         Renderer.act(() => {
    //             rnInput.props.onChangeText('Hello!')
    //         })

    //         expect(rnInput.props.value).toStrictEqual('Hello!')
    //     })
    // })

    describe('Snapshots', () => {
        it('render run with default props', () => {
            const mockFunc = jest.fn()

            const renderer = new ShallowRenderer();
            const output = renderer.render(<Input onChangeText={mockFunc} value='value'/>);

            expect(output).toMatchSnapshot();
        })

        it('render run with some props', () => {
            const mockFunc = jest.fn()

            const renderer = new ShallowRenderer();
            const output = renderer.render(
                <Input
                    onChangeText={mockFunc}
                    value='value'
                    placeholder='placeholder'
                    type='password'
                />
            );

            expect(output).toMatchSnapshot();
        })
    })
})