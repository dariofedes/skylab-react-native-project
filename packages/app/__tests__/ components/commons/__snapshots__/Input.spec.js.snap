// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component Snapshots render run with default props 1`] = `
<Component
  allowFontScaling={true}
  keyboardType="default"
  onChangeText={[MockFunction]}
  placeholder="Type here..."
  placeholderTextColor="#666"
  rejectResponderTermination={true}
  secureTextEntry={false}
  style={
    Object {
      "borderColor": "#333",
      "borderRadius": 10,
      "borderWidth": 1,
      "color": "#333",
      "margin": 10,
      "paddingHorizontal": 10,
      "paddingVertical": 10,
    }
  }
  underlineColorAndroid="transparent"
  value="value"
/>
`;

exports[`Input Component Snapshots render run with some props 1`] = `
<Component
  allowFontScaling={true}
  keyboardType="default"
  onChangeText={[MockFunction]}
  placeholder="placeholder"
  placeholderTextColor="#666"
  rejectResponderTermination={true}
  secureTextEntry={true}
  style={
    Object {
      "borderColor": "#333",
      "borderRadius": 10,
      "borderWidth": 1,
      "color": "#333",
      "margin": 10,
      "paddingHorizontal": 10,
      "paddingVertical": 10,
    }
  }
  underlineColorAndroid="transparent"
  value="value"
/>
`;
